src/mscorlib/src/System/IO/Stream.cs

```diff
@@ -794,15 +793,14 @@ public virtual void Write(ReadOnlySpan<byte> source)
         }
 
         // Writes one byte from the stream by calling Write(byte[], int, int).
-        // This implementation does not perform well because it allocates a new
-        // byte[] each time you call it, and should be overridden by any 
+        // This implementation should be overridden by any 
         // subclass that maintains an internal buffer.  Then, it can help perf
         // significantly for people who are writing one byte at a time.
         public virtual void WriteByte(byte value)
         {
-            byte[] oneByteArray = new byte[1];
-            oneByteArray[0] = value;
-            Write(oneByteArray, 0, 1);
+            Span<byte> bytes = stackalloc byte[1];
```

Comment:
Do you even need to stackalloc here?

```C#
public virtual unsafe WriteByte(byte value)
{
   Write(new Span<byte>(&value, 1));
}
```

it's unsafe though....

Reply:
Didn't think it would make much difference with the improvements to small fixed stackalloc https://github.com/dotnet/coreclr/pull/14623; but its outrageously better

ReadByte with
```csharp
Span<byte> bytes = stackalloc byte[1];
```
Produces
```asm
; Lcl frame size = 112
G_M10894_IG01:
       push     rbp
       push     rdi
       push     rsi
       sub      rsp, 112
       lea      rbp, [rsp+20H]
       mov      rsi, rcx
       lea      rdi, [rbp+10H]
       mov      ecx, 14
       xor      rax, rax
       rep stosd 
       mov      rcx, rsi
       mov      qword ptr [rbp+48H], rsp
       mov      rax, qword ptr [(reloc)]
       mov      qword ptr [rbp+08H], rax
G_M10894_IG02:
       lea      rdx, bword ptr [rbp+28H]
G_M10894_IG03:
       xorpd    xmm0, xmm0
       movdqu   qword ptr [rdx], xmm0
G_M10894_IG04:
       add      rsp, 32
       push     0
       push     0
       sub      rsp, 32
       lea      rdx, [rsp+20H]
       mov      qword ptr [rbp+48H], rsp
       xor      rax, rax
       mov      qword ptr [rbp+20H], rax
       mov      bword ptr [rbp+20H], rdx
       mov      rdx, bword ptr [rbp+20H]
       lea      rax, bword ptr [rbp+28H]
       mov      bword ptr [rax], rdx
       mov      dword ptr [rbp+30H], 1
G_M10894_IG05:
       movdqu   xmm0, qword ptr [rbp+28H]
       movdqu   qword ptr [rbp+38H], xmm0
G_M10894_IG06:
       movdqu   xmm0, qword ptr [rbp+38H]
       movdqu   qword ptr [rbp+10H], xmm0
G_M10894_IG07:
       lea      rdx, bword ptr [rbp+10H]
       mov      rax, qword ptr [rcx]
       mov      rax, qword ptr [rax+96]
       call     qword ptr [rax+24]Stream:Read(struct):int:this
       test     eax, eax
       jne      SHORT G_M10894_IG10
       mov      eax, -1
       lea      rcx, [(reloc)]
       mov      rcx, qword ptr [rcx]
       cmp      qword ptr [rbp+08H], rcx
       je       SHORT G_M10894_IG08
       call     CORINFO_HELP_FAIL_FAST
G_M10894_IG08:
       nop      
G_M10894_IG09:
       lea      rsp, [rbp+50H]
       pop      rsi
       pop      rdi
       pop      rbp
       ret      
G_M10894_IG10:
       cmp      dword ptr [rbp+40H], 0
       jbe      SHORT G_M10894_IG13
       mov      rax, bword ptr [rbp+38H]
       movzx    rax, byte  ptr [rax]
       lea      rcx, [(reloc)]
       mov      rcx, qword ptr [rcx]
       cmp      qword ptr [rbp+08H], rcx
       je       SHORT G_M10894_IG11
       call     CORINFO_HELP_FAIL_FAST
G_M10894_IG11:
       nop      
G_M10894_IG12:
       lea      rsp, [rbp+50H]
       pop      rsi
       pop      rdi
       pop      rbp
       ret      
G_M10894_IG13:
       call     CORINFO_HELP_RNGCHKFAIL
       int3     
; Total bytes of code 229, prolog size 46 for method Stream:ReadByte():int:this
```
Changing the one line to these two lines 
```csharp
byte b = 0;
Span<byte> bytes = new Span<byte>(ref b, 1);
```
Produces
```asm
; Lcl frame size = 64
G_M10891_IG01:
       push     rsi
       sub      rsp, 64
       xor      rax, rax
       mov      qword ptr [rsp+28H], rax
G_M10891_IG02:
       xor      edx, edx
       mov      dword ptr [rsp+3CH], edx
       lea      rdx, bword ptr [rsp+3CH]
       mov      rsi, rdx
       mov      rdx, rsi
       mov      eax, 1
       mov      rsi, rdx
       lea      rdx, bword ptr [rsp+28H]
       mov      bword ptr [rdx], rsi
       mov      dword ptr [rdx+8], eax
       lea      rdx, bword ptr [rsp+28H]
       mov      rax, qword ptr [rcx]
       mov      rax, qword ptr [rax+96]
       call     qword ptr [rax+24]Stream:Read(struct):int:this
       test     eax, eax
       jne      SHORT G_M10891_IG04
       mov      eax, -1
G_M10891_IG03:
       add      rsp, 64
       pop      rsi
       ret      
G_M10891_IG04:
       movzx    rax, byte  ptr [rsi]
G_M10891_IG05:
       add      rsp, 64
       pop      rsi
       ret      
; Total bytes of code 87, prolog size 12 for method Stream:ReadByte():int:this
```

Reply:
Hmm, that does look strange; I would not expect them to be all that different. 

I'll take a look at what is going on under the hood.

Reply:
Looks like csc is storing the length in a local and so blocks the jit's ability to convert the stackalloc as it doesn't see that it is a fixed size.
```
  IL_0000:  ldc.i4.1
  IL_0001:  stloc.1
  IL_0002:  ldloc.1
  IL_0003:  conv.u
  IL_0004:  localloc
```
The localloc transformation happens early before we've set up any kind of dataflow capabilities in the jit. So despite this looking like something that should trivially fall out we don't have the mechansims to catch this today.

Reply:
If I hack the jit to forcibly enable the fixed stackalloc transformation then `Unsafe.As` fails to inline in a manner similar to #11211 -- here the argument is the address of a local and so arg fetching bypasses the fix from #11218.

With the unconverted stackalloc `Unsafe.As` inlines ok.